### PR TITLE
Feat/e rosita commissioning support

### PR DESCRIPTION
**Pull request by jessicapilling**

### CODE REVIEW

File: daxa/archive/base.py
Comment (DavidT3) [inline daxa/archive/base.py:47]: Please correct my spelling mistake here, missions class instance -> mission class instance

File: daxa/config.py
Comment (DavidT3) [inline daxa/config.py:75]: CalPV_info should be all uppercase, that is the Python PEP convention for constants, it distinguishes them from normal variables (though not in their behaviour)

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:398]: Similar deal here, make use of a with statement, then you won't need tar.close() at the end.

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:418]: Informative variable names please. Single letter variables are to be avoided, apart from perhaps in list comprehensions. This could be 'for obs_id in self.filtered_obs_ids:'

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:419]: This format is not necessary, just do self.raw_data_path + obs_id

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:422]: This format is not necessary, just do self.raw_data_path + obs_id

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:428]: I didn't know about startswith, that's cool!

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:444]: What files are you deleting? Some of them might be necessary for further processing and cleaning, although you know better than me on that front.

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:459]: That format is not necessary, just do self.raw_data_path + obs.

Though also it would be better if that changed to obs_id, that's generally that name I used for an ObsID variable (though I'm definitely not super consistent in doing that).

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:463]: Are there any other ways to tell the files apart?

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:465]: Same comment about the format thing

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:472]: It could be tired David brain but I'm not sure the Fields sentence actually makes sense

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:475]: Update comment to say eROSITA Calibration or something, rather than XMM

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:477]: You could use your raw data directory property here

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:500]: Same comment about the format thing with obs id

Comment (DavidT3) [reply daxa/mission/erosita.py:500]: Also, I think this sort of thing will become much simpler once you put all the information about observations and fields into one file, as you'll just be able to grab the fields column of the filtered_obs_info property and that will just tell you which fields have to be acquired.

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:508]: Remember to remove self if you make _download_call static again like I suggested.

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:512]: I'm afraid this chunk just fully doesn't work - it gives errors about there being no self passed to the _download_call method. That will go away once you make it static again and then as far as I can tell it will work

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:549]: Yeah so this is what causes the issue, as you don't pass self - again won't be necessary once download_call is static

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:572]: Setting up the progress bar with the variable name 'download_prog' doesn't really make sense here as you aren't downloading.

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:584]: Setting up the progress bar with the variable name 'download_prog' doesn't really make sense here as you aren't downloading.

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:594]: If you rename the progress bar variable this will have to change.

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:604]: If you rename the progress bar variable this will have to change.

File: daxa/files/obs_info.csv
Comment (DavidT3) [reply daxa/files/obs_info.csv:1]: Also, this file has an index column (the , at the beginning of the column names, and the 0, 1, 2, 3, 4 etc. that is the first entry. That's unnecessary and appears as an unsightly named column when you look at the all_obs_info dictionary.

image

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:246]: Before you send it to the all_obs_info property it would be nice to re-order the columns. One way of doing it is this:

https://github.com/DavidT3/DAXA/blob/a04557d0ba2b2ba5eaf7da56b7b2418cfb714ff2/daxa/mission/nustar.py#L278-L281

Though if you forget to put a column in the list of column names it will be removed from the dataframe so just be wary of that. 

I just think it's nicer to have the ra, dec, and ObsID being the first things the user sees on the left - entirely aesthetic decision though

**Review by DavidT3 — CHANGES_REQUESTED** [review]
So, I have finished the first review I've ever done on GitHub (which was eye opening for me). You've done a good job getting this mission class done, and while I've put quite a few comments in, a lot of them are minor things.

The largest change I want is for you to change the way you store information about the eROSITA observations - having two files is a detriment, I think they can be combined so that on each row is an ObsID and then the information like field name and field type and field download link is just there for each ObsID. It duplicates information slightly but that hardly matters for a ~160 line csv.

It will make a lot of the class much cleaner, because you'll just be able to grab the filtered_obs_info dataframe and look at the field column, get the unique values of that column, and oh that's what you need to download.

Anyway, you should be able to see all of my comments I hope, and feel free to email me about them/berate me.

You've done a great job!

**Review by DavidT3 — CHANGES_REQUESTED** [review]
So, I have finished the first review I've ever done on GitHub (which was eye opening for me). You've done a good job getting this mission class done, and while I've put quite a few comments in, a lot of them are minor things.

The largest change I want is for you to change the way you store information about the eROSITA observations - having two files is a detriment, I think they can be combined so that on each row is an ObsID and then the information like field name and field type and field download link is just there for each ObsID. It duplicates information slightly but that hardly matters for a ~160 line csv.

It will make a lot of the class much cleaner, because you'll just be able to grab the filtered_obs_info dataframe and look at the field column, get the unique values of that column, and oh that's what you need to download.

Anyway, you should be able to see all of my comments I hope, and feel free to email me about them/berate me.

You've done a great job!

File: daxa/config.py
Comment (jessicapilling) [reply daxa/config.py:76]: No reason, I will make them into one :)

File: daxa/files/eROSITACalPV_info.csv
Comment (jessicapilling) [reply daxa/files/eROSITACalPV_info.csv]: Completely agree, see the reply to your comment above.

File: daxa/files/obs_info.csv
Comment (jessicapilling) [reply daxa/files/obs_info.csv]: Unsightly common is gone in the new merged csv

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py]: deleted this function when the two csv's were merged

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py:269]: I testing it by trying eROSITACalPV(fields='crab').chosen_fields, which returns: ['CRAB_1', 'CRAB_2', 'CRAB_4', 'CRAB_3'].  Not sure why the CRAB_3 and CRAB_4 are out of order though.

Comment (jessicapilling) [reply daxa/mission/erosita.py]: Ok no matter anyway since I changed how the _check_chos_fields() function reads in the fields argument

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py]: this is fine, I have updated this function so there is no need for indexing

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py]: Ah! After the first comment on this function I went and changed it to work by a case by case basis so that was nice we had the same thinking! It is MUCH easier to read now as well :)

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py:308]: I realised I also never updated the docstring param to explain evlist_path either, so I've updated that.

The [-:5] removes the '.fits' part of the file path string, so that it can be replaced with the ending that gets added to the file name after instrument filtering.

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py]: just checking that this is about just line 372?

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py]: hstack is turning the list comprehension into an array so it can be used to index the data

the .nonzero() turns the (t_col==i) from an array of Bools into the indexes of t_col==i is True

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py]: Added this to the _download_call() and _directory_formatting function

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py:432]: I could also do: file_name = [file for file in all_files if 'if' not in file and not file.startswith('.')][0]

so that gets rid of the instrument filtered files with the 'if' suffix and the hidden files

Do you think that is more readable?

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py:516]: yeah, all these comments I think are fully resolved, I can run the download() function after I made it a proper static method

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py]: yup, changed the name

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py]: thats all done now

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py:228]: no worries, I've put this in the updated version of _fetch_obs_info() when I merged the two csv's

File: daxa/mission/erosita.py
Comment (DavidT3) [reply daxa/mission/erosita.py]: Would you like the satisfaction of closing issue 108?

File: daxa/mission/erosita.py
Comment (DavidT3) [reply daxa/mission/erosita.py]: > I testing it by trying eROSITACalPV(fields='crab').chosen_fields, which returns: ['CRAB_1', 'CRAB_2', 'CRAB_4', 'CRAB_3']. Not sure why the CRAB_3 and CRAB_4 are out of order though.

My point was more, what if somebody passes ['CRAB', 'eFEDS'], then I think slicing like that would overwrite eFEDS. Anyway doesn't matter if you've changed it - I might comment more in the second review.

File: daxa/mission/erosita.py
Comment (DavidT3) [reply daxa/mission/erosita.py:312]: > I realised I also never updated the docstring param to explain evlist_path either, so I've updated that.
> 
> The [-:5] removes the '.fits' part of the file path string, so that it can be replaced with the ending that gets added to the file name after instrument filtering.

Ah I see of course! Another way of doing that might be evlist_path.split('.')[0] or split('.fits')[0] - not requiring you to change just an FYI

File: daxa/mission/erosita.py
Comment (DavidT3) [reply daxa/mission/erosita.py]: I think np.where might be a better way of doing this - list comprehensions are going to be sloooow here I would think.

File: daxa/mission/erosita.py
Comment (DavidT3) [reply daxa/mission/erosita.py]: Ah yes sorry for not making that clear

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:484]: When I deleted all of the data I acquired the first time I tested this, and tried to download on multiple cores again, it did do the download but afterwards an index error got thrown.

File: daxa/mission/erosita.py
Comment (DavidT3) [inline daxa/mission/erosita.py:314]: I noticed the first time I reviewed the pull request (but forgot to mention) that if data are already downloaded (cached I guess we would call it), and then another run of eROSITACalPV is done but with instruments filtered, the requested event lists aren't created.

**Review by DavidT3 — CHANGES_REQUESTED** [review]
Very nearly there now! Just a few more changes and this can go into the master branch.

One thing I have thought of is that the whole 'filtering on positions' thing won't really be working for survey fields of the CalPV stuff, but I think that will have to be okay for now.

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py]: I timed it and its like 7 times faster to do np.where

File: daxa/mission/base.py
Comment (jessicapilling) [reply daxa/mission/base.py:575]: Would you like me to add in that I've overwritten this from base.py in the docstring? I thought because its a user facing method its best in the comments

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py:516]: this was an issue with the _directory_formatting function.

When I changed the _download_call to have field specific download folders instead of having them all in temp_download, there was another layer of folders that _directory_formatting couldnt deal with, so I have added in another check for the extra layer.

This is resolved in the commit that mentions _directory_formating

File: daxa/mission/erosita.py
Comment (jessicapilling) [reply daxa/mission/erosita.py:340]: Ahh I see what you mean, it will only filter instruments when the download() is called at the moment, I will change how _inst_filtering() works, and then call it earlier

Comment (jessicapilling) [reply daxa/mission/erosita.py:343]: ok i had a thought, 

since I parallelised the inst_filtering function, shouldnt it be a static method like _download_call?

Then if I want to do that, I will leave the inst_filtering function as it was (except now static), and instead overwrite the chosen_instruments function

Comment (jessicapilling) [reply daxa/mission/erosita.py:447]: Ok cool, I've committed this change, if you see this before monday I haven't tested it yet, so be careful!

**Comment by DavidT3** [comment]
This pull request cannot be completed as we've got slightly turned around somewhere, and the commit history of this branch would now wipe out all changes made to master after the creation of this branch.

I think this is because (for some reason) master was merged into the dev branch on the remote, but somehow in the meantime several commits were made locally on the dev's machine. That local branch hadn't had master merged in, so the eventual differences means that all master additions would be deleted.

Anyway, I have fixed this with a new branch (feat/fixCommiteROSITACalPVSupport) by resetting to before the merge commit, then cherry picking all the subsequent commits. A new PR for that branch will be created, and this one will be referenced.

**Pull request closed without merging**
